fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneThenKeyExpression.java

```diff
@@ -0,0 +1,141 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+public class LuceneThenKeyExpression extends ThenKeyExpression implements LuceneKeyExpression {
+
+    // Only one key allowed per repeated field
+    private KeyExpression primaryKey;
+    boolean validated = false;
+
+
+    public LuceneThenKeyExpression(@Nullable final KeyExpression primaryKey, @Nonnull final List<KeyExpression> children) {
+        super(children);
+        if (!validate()) {
+            throw new IllegalArgumentException("failed to validate lucene compatibility on construction");
+        }
+        this.primaryKey = primaryKey;
+
+    }
+
+    // Todo: figure out if we should allow stored key expressions inside a thenKeyExpression,
+    // for now not but tbd if we should check the farOut and decide based on that.
+    public boolean validate() {
+        if (validated) {
+            return validated;
+        }
+        validated = true;
+        for (KeyExpression child : getChildren()) {
+            validated = validated && (primaryKey != null) ? child instanceof LuceneFieldKeyExpression : (LuceneKeyExpression.validateLucene(child));
+        }
+        return validated;
+    }
+
+    public KeyExpression getPrimaryKey() {
+        return primaryKey;
+    }
+
+    public int getPrimaryKeyPosition() {
+        return super.normalizeKeyForPositions().indexOf(primaryKey);
+    }
+
+    public List<LuceneFieldKeyExpression> getLuceneChildren() {
+        List<LuceneFieldKeyExpression> children = new LinkedList<>();
+        KeyExpressionVisitor visitor = new KeyExpressionVisitor() {
+
+            @Override
+            public KeyExpression visitField(final FieldKeyExpression fke) {
+                if (fke instanceof LuceneFieldKeyExpression) {
+                    children.add((LuceneFieldKeyExpression)fke);
+                }
+                return fke;
+            }
+
+            @Override
+            public KeyExpression visitThen(final ThenKeyExpression thenKey) {
+                for (KeyExpression child : thenKey.getChildren()) {
+                    visit(child);
+                }
+                return thenKey;
+            }
+
+            @Override
+            public KeyExpression visitNestingKey(final NestingKeyExpression nke) {
+                visitField(nke.getParent());
+                visit(nke.getChild());
+                return nke;
+            }
+
+            @Override
+            public KeyExpression visitGroupingKey(final GroupingKeyExpression gke) {
+                visit(gke.getChild());
+                return gke;
+            }
+
+            @Override
+            public boolean applies(final String indexType) {
+                return IndexTypes.LUCENE.equals(indexType);
+            }
+        };
+
+        visitor.visit(this);
+
+        return children;
+    }
+
+    public boolean fan() {
```

Comment:
This method needs a comment; it feels to me like the sense is backwards: a primary-key then-field doesn't fan out into multiple document fields.

Reply:
If it has a primary key then it does fan out repeatedly and that primary key value is used to prefix it. I think this is consistent with how we use Fan-Out right?

Reply:
Yes, you are right. I think it still needs a comment explaining that this means that one of the children is a key-like field that gets made part of the full field name, if that's what it does.